### PR TITLE
[MI-1917]: Add radio field and input field components

### PR DESCRIPTION
**Pull request by ayusht2810**
Add radio field and input field components to reuse them in other components.

### CODE REVIEW

File: webapp/src/components/InputField/InputField.tsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    };
+    description: {
+        id: string;
+        message: string;
+    }
+}
+
+export const InputField = (props: Props) => {
+    console.log(props.input.defaultValue);
+    console.log(props.tagType);
```
Comment (raghavaggarwal2308) [inline webapp/src/components/InputField/InputField.tsx]: remove extra console statements

File: webapp/src/components/InputField/InputField.tsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        value: string | number;
+        disabled?: boolean;
+    };
+    description: {
+        id: string;
+        message: string;
+    }
+}
+
+export const InputField = (props: Props) => {
```
Comment (raghavaggarwal2308) [inline webapp/src/components/InputField/InputField.tsx]: destructure props

File: webapp/src/components/InputField/InputField.tsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    };
+    tagType: any;
+    input: {
+        type?: string;
+        placeholder?: string;
+        maxLength?: number;
+        min?: number;
+        rows?: number;
+        defaultValue?: number;
+        onChange: (e: any) => void;
```
Comment (raghavaggarwal2308) [inline webapp/src/components/InputField/InputField.tsx]: use proper types

File: webapp/src/components/InputField/InputField.tsx [hunk 4]
```diff
@@ -0,0 +1,64 @@
+import * as React from 'react';
+import {FormattedMessage} from 'react-intl';
+
+type Props = {
+    heading: {
+        id: string;
+        message: string;
+    };
+    tagType: any;
```
Comment (raghavaggarwal2308) [inline webapp/src/components/InputField/InputField.tsx]: use proper type

File: webapp/src/components/admin_settings/jaas_section.tsx [hunk 5]
```diff
@@ -16,158 +18,115 @@ type Props = {
...
+                }}
+                labelClass={'radio-inline'}
+                options={EMBED_JAAS_VIDEO_INSIDE_MATTERMOST_OPTIONS}
+                onChange={props.onEmbeddedChange}
+                description={{
+                    id: 'jaas.embed-meeting-description',
+                    message: '(Experimental) When true, JaaS video is embedded as a floating window inside Mattermost by default.'
+                }}
+            />
+
```
Comment (raghavaggarwal2308) [inline webapp/src/components/admin_settings/jaas_section.tsx]:
```suggestion
```

File: webapp/src/components/admin_settings/jaas_section.tsx [hunk 6]
```diff
@@ -16,158 +18,115 @@ type Props = {
...
+                }}
+                labelClass={'radio-inline'}
+                options={JAAS_COMPATIBILITY_OPTIONS}
+                onChange={props.onCompatibilityChange}
+                description={{
+                    id: 'jaas.enable-compatibility-mode-description',
+                    message: '(Insecure) If your JaaS server is not compatible with this plugin, include the JavaScript API hosted on your Jaas server directly in Mattermost instead of the default API version provided by the plugin. WARNING: Enabling this setting can compromise the security of your Mattermost system, if your JaaS server is not fully trusted and allows direct modification of program files. Use with caution.'
+                }}
+            />
+
```
Comment (raghavaggarwal2308) [inline webapp/src/components/admin_settings/jaas_section.tsx]:
```suggestion
```

File: webapp/src/components/admin_settings/jaas_section.tsx [hunk 7]
```diff
@@ -16,158 +18,115 @@ type Props = {
...
+                    onChange: props.onAppIDChange,
+                    value: props.appID,
+                    disabled: props.disabled
+                }}
+                description={{
+                    id: 'jaas.app-id-description',
+                    message: 'Specify your JaaS AppID. You can get the AppID from https://jaas.8x8.vc/#/apikeys.'
+                }}
+            />
+
```
Comment (raghavaggarwal2308) [inline webapp/src/components/admin_settings/jaas_section.tsx]:
```suggestion
```

File: webapp/src/components/admin_settings/jaas_section.tsx [hunk 8]
```diff
@@ -16,158 +18,115 @@ type Props = {
...
+                    onChange: props.onApiKeyIDChange,
+                    value: props.apiKey,
+                    disabled: props.disabled
+                }}
+                description={{
+                    id: 'jaas.api-key-id-description',
+                    message: 'Specify your JaaS Api key ID. You can get the Api key ID from https://jaas.8x8.vc/#/apikeys.'
+                }}
+            />
+
```
Comment (raghavaggarwal2308) [inline webapp/src/components/admin_settings/jaas_section.tsx]:
```suggestion
```

File: webapp/src/components/admin_settings/jaas_section.tsx [hunk 9]
```diff
@@ -16,158 +18,115 @@ type Props = {
...
+                    onChange: props.onPrivateKeyChange,
+                    value: props.privateKey,
+                    disabled: props.disabled
+                }}
+                description={{
+                    id: 'jaas.rsa-key-description',
+                    message: 'Specify your JaaS private key. You can get the private key by generating an API key pair from https://jaas.8x8.vc/#/apikeys OR from https://jaas.8x8.vc/#/start-guide in the "Api Key" section. (NOTE: While generating the key manually make sure to set the size to be 2048 and not 4096).'
+                }}
+            />
+
```
Comment (raghavaggarwal2308) [inline webapp/src/components/admin_settings/jaas_section.tsx]:
```suggestion
```

File: webapp/src/components/admin_settings/jitsi_section.tsx [hunk 10]
```diff
@@ -29,293 +31,199 @@ export const JITSI_NAMING_SCHEME = {
...
+                }}
+                labelClass={'radio-inline'}
+                options={JITSI_COMPATIBILITY_OPTIONS}
+                onChange={props.onJitsiCompatibilityChange}
+                description={{
+                    id: 'jitsi.enable-compatibility-mode-description',
+                    message: '(Insecure) If your Jitsi server is not compatible with this plugin, include the JavaScript API hosted on your Jitsi server directly in Mattermost instead of the default API version provided by the plugin. WARNING: Enabling this setting can compromise the security of your Mattermost system, if your Jitsi server is not fully trusted and allows direct modification of program files. Use with caution.'
+                }}
+            />
+
```
Comment (raghavaggarwal2308) [inline webapp/src/components/admin_settings/jitsi_section.tsx]:
```suggestion
```

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: webapp/src/components/InputField/InputField.tsx [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    };
+    tagType: string;
+    input: {
+        type?: string;
+        placeholder?: string;
+        maxLength?: number;
+        min?: number;
+        rows?: number;
+        defaultValue?: number;
+        onChange: (e: React.ChangeEvent<any>) => void;
```
Comment (raghavaggarwal2308) [inline webapp/src/components/InputField/InputField.tsx]: remove any and try to use proper type

File: webapp/src/components/RadioField/RadioField.tsx [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+                                <>
+                                    <label className={`pt-0 ${labelClass}`}>
+                                        <input
+                                            key={option.value}
+                                            type='radio'
+                                            value={option.value}
+                                            checked={option.checked}
+                                            onChange={onChange}
+                                        />
+                                        {option.value === 'true' || option.value === 'false' ?
```
Comment (raghavaggarwal2308) [inline webapp/src/components/RadioField/RadioField.tsx]: remove this condition and add translation to "true" and "false" also.....I guess it was missed by me.

File: webapp/src/components/admin_settings/jaas_section.tsx [hunk 13]
```diff
@@ -1,5 +1,7 @@
 import * as React from 'react';
-import {FormattedMessage} from 'react-intl';
+
+import {InputField} from '../InputField/InputField';
+import {RadioField} from '../RadioField/RadioField';
```
Comment (raghavaggarwal2308) [inline webapp/src/components/admin_settings/jaas_section.tsx]: rename these files to `index.tsx` then u can specify the path as `../InputField`......that looks neater than writing `../InputField/InputFiel`

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: webapp/src/components/InputField/InputField.tsx (hunk 11 above)
Comment (ayusht2810) [reply webapp/src/components/InputField/InputField.tsx]: used "Element" type (base of all elements) was still getting TS error

File: webapp/src/components/RadioField/index.tsx [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+            <div className='form-group'>
+                <label className='col-sm-4'>
+                    <FormattedMessage
+                        id={heading.id}
+                        defaultMessage={heading.message}
+                    />
+                </label>
+                <div className='col-sm-8'>
+                    <>
+                        <div className={`${divClass}`}>
```
Comment (raghavaggarwal2308) [inline webapp/src/components/RadioField/index.tsx]: use meaningful name instead of `divClass`

**Review by raghavaggarwal2308 — APPROVED** [review]

File: webapp/i18n/en.json [hunk 15]
```diff
@@ -7,6 +7,10 @@
   "jaas.embed-meeting-description": "(Experimental) When true, JaaS video is embedded as a floating window inside Mattermost by default.",
   "jaas.rsa-key": "RSA Private key for JaaS Authentication:",
   "jaas.rsa-key-description": "Specify your JaaS private key. You can get the private key by generating an API key pair from https://jaas.8x8.vc/#/apikeys OR from https://jaas.8x8.vc/#/start-guide in the \"Api Key\" section. (NOTE: While generating the key manually make sure to set the size to be 2048 and not 4096).",
+  "jaas.embed-enable": "true",
+  "jaas.embed-disable": "false",
```
Comment (manojmalik20) [inline webapp/i18n/en.json]: There should only be two ids and these should be used everywhere.
```suggestion
  "jitsi.true": "true",
  "jitsi.false": "false",
```

File: webapp/src/components/InputField/index.tsx [hunk 16]
```diff
@@ -0,0 +1,74 @@
+import * as React from 'react';
+import {FormattedMessage} from 'react-intl';
+
+import {InputFieldType} from 'types';
+
+type Props = {
+    heading: {
+        id: string;
+        message: string;
+    };
```
Comment (manojmalik20) [inline webapp/src/components/InputField/index.tsx]:
```suggestion
    heading: string | ReactNode;
```
Refer https://github.com/mattermost/mattermost-webapp/blob/7c09150efdca6e488efedadea793b2b681822a48/components/quick_input/quick_input.tsx#L43

File: webapp/src/components/InputField/index.tsx [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+        rows?: number;
+        defaultValue?: number;
+        onChange: (e: React.ChangeEvent<InputFieldType>) => void;
+        value: string | number;
+        disabled?: boolean;
+    };
+    description: {
+        id: string;
+        message: string;
+    }
```
Comment (manojmalik20) [inline webapp/src/components/InputField/index.tsx]: Same.

File: webapp/src/components/InputField/index.tsx [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+    input: {
+        type?: string;
+        placeholder?: string;
+        maxLength?: number;
+        min?: number;
+        rows?: number;
+        defaultValue?: number;
+        onChange: (e: React.ChangeEvent<InputFieldType>) => void;
+        value: string | number;
+        disabled?: boolean;
```
Comment (manojmalik20) [inline webapp/src/components/InputField/index.tsx]: Remove these fields from inside the `input` key. These should be the direct keys of the props.

File: webapp/src/components/InputField/index.tsx [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+        <>
+            <div className='form-group'>
+                <label className='col-sm-4'>
+                    <FormattedMessage
+                        id={heading.id}
+                        defaultMessage={heading.message}
+                    />
+                </label>
+                <div className='col-sm-8'>
+                    {tagType === 'input' ? (
```
Comment (manojmalik20) [inline webapp/src/components/InputField/index.tsx]: There's no need for this `tagType` if you are getting the type in the`type` prop.

File: webapp/src/components/InputField/index.tsx [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+        value: string | number;
+        disabled?: boolean;
+    };
+    description: {
+        id: string;
+        message: string;
+    }
+}
+
+export const InputField = ({heading, tagType, input, description}: Props) => {
```
Comment (manojmalik20) [inline webapp/src/components/InputField/index.tsx]:
```suggestion
export const TextInput = ({heading, tagType, input, description}: Props) => {
```

File: webapp/src/components/RadioField/index.tsx [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+    onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
+    description: {
+        id: string;
+        message: string;
+    }
+}
+
+export const RadioField = ({heading, radioGlobal, radioInline, options, onChange, description}: Props) => {
+    return (
+        <>
```
Comment (manojmalik20) [inline webapp/src/components/RadioField/index.tsx]: I don't think there's any need for this here.

File: webapp/src/components/RadioField/index.tsx [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+                    <FormattedMessage
+                        id={heading.id}
+                        defaultMessage={heading.message}
+                    />
+                </label>
+                <div className='col-sm-8'>
+                    <>
+                        <div className={`${radioGlobal}`}>
+                            {options.map((option) => (
+                                <>
```
Comment (manojmalik20) [inline webapp/src/components/RadioField/index.tsx]: Same here. Remove unnecessary use of these from everywhere.

File: webapp/src/components/RadioField/index.tsx [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+        message?: string;
+    }[];
+    onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
+    description: {
+        id: string;
+        message: string;
+    }
+}
+
+export const RadioField = ({heading, radioGlobal, radioInline, options, onChange, description}: Props) => {
```
Comment (manojmalik20) [inline webapp/src/components/RadioField/index.tsx]:
```suggestion
export const RadioField = ({heading, radioGlobal = '', radioInline = '', options, onChange, description}: Props) => {
```

File: webapp/src/components/RadioField/index.tsx [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+                </div>
+            </div>
+        </>
+    );
+};
+
+RadioField.defaultProps = {
+    radioGlobal: '',
+    radioInline: ''
+};
```
Comment (manojmalik20) [inline webapp/src/components/RadioField/index.tsx]:
```suggestion
```

File: webapp/src/components/RadioField/index.tsx [hunk 25]
```diff
@@ -0,0 +1,73 @@
+import * as React from 'react';
+import {FormattedMessage} from 'react-intl';
+
+type Props = {
+    heading: {
+        id: string;
+        message: string;
+    };
+    radioGlobal?: string;
+    radioInline?: string;
```
Comment (manojmalik20) [inline webapp/src/components/RadioField/index.tsx]:
```suggestion
    radioGlobalClassName?: string;
    radioInlineClassName?: string;
```

File: webapp/src/components/RadioField/index.tsx [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+    options: {
+        value: string;
+        checked: boolean;
+        id?: string;
+        message?: string;
```
Comment (manojmalik20) [inline webapp/src/components/RadioField/index.tsx]: You should create a type for this.

File: webapp/src/components/admin_settings/jaas_section.tsx [hunk 27]
```diff
@@ -1,170 +1,136 @@
...
+            id: 'jaas.compatibility-enable',
+            message: 'true'
+        },
+        {
+            value: 'false',
+            checked: !compatibilityMode,
+            id: 'jaas.compatibility-disable',
+            message: 'false'
+        }
+    ];
```
Comment (manojmalik20) [inline webapp/src/components/admin_settings/jaas_section.tsx]: Use `useMemo`.

File: webapp/src/components/admin_settings/jaas_section.tsx [hunk 28]
```diff
@@ -1,170 +1,136 @@
...
-                        </span>
-                    </div>
-                </div>
-            </div>
+            <RadioField
+                heading={{
+                    id: 'jaas.embed-meeting',
+                    message: 'Embed JaaS video inside Mattermost:'
+                }}
+                radioInline={'radio-inline'}
```
Comment (manojmalik20) [inline webapp/src/components/admin_settings/jaas_section.tsx]: I think this should be a boolean prop `isInline`

File: webapp/src/components/admin_settings/jaas_section.tsx [hunk 29]
```diff
@@ -1,170 +1,136 @@
...
+            />
+            <InputField
+                heading={{
+                    id: 'jaas.app-id',
+                    message: 'AppID for JaaS Authentication:'
+                }}
+                tagType={'input'}
+                input={{
+                    type: 'input',
+                    maxLength: -1,
```
Comment (manojmalik20) [inline webapp/src/components/admin_settings/jaas_section.tsx]: Is there any need for this?

File: webapp/src/components/admin_settings/jitsi_settings.tsx [hunk 30]
```diff
@@ -77,6 +78,21 @@ const JitsiSettings = ({id, value, disabled, config, onChange, setSaveNeeded}: P
...
+            id: 'jitsi.input-enable-jitsi',
+            message: 'Jitsi'
+        },
+        {
+            value: JAAS_MODE,
+            checked: mode === JAAS_MODE,
+            id: 'jitsi.input-enable-jass',
+            message: 'JasS'
+        }
+    ];
```
Comment (manojmalik20) [inline webapp/src/components/admin_settings/jitsi_settings.tsx]: Use `useMemo`.

File: webapp/src/components/admin_settings/jaas_section.tsx [hunk 31]
```diff
@@ -1,170 +1,136 @@
...
+import {RadioField} from '../RadioField';
 
 type Props = {
-    disabled: boolean,
-    onAppIDChange: (e: React.ChangeEvent<HTMLInputElement>) => void,
-    onApiKeyIDChange: (e: React.ChangeEvent<HTMLInputElement>) => void,
-    onPrivateKeyChange: (e: React.ChangeEvent<HTMLTextAreaElement>) => void,
-    onEmbeddedChange: (e: React.ChangeEvent<HTMLInputElement>) => void,
-    onCompatibilityChange: (e: React.ChangeEvent<HTMLInputElement>) => void,
+    disable: boolean,
```
Comment (manojmalik20) [inline webapp/src/components/admin_settings/jaas_section.tsx]: Why this change?

File: webapp/src/components/admin_settings/jitsi_section.tsx [hunk 32]
```diff
@@ -29,290 +32,202 @@ export const JITSI_NAMING_SCHEME = {
...
+                    id: 'jitsi.embed-video-description',
+                    message: '(Experimental) When true, Jitsi video is embedded as a floating window inside Mattermost by default. Users can override this setting with \'/jitsi settings\'.'
+                }}
+            />
+            <RadioField
+                heading={{
+                    id: 'jitsi.meeting-names',
+                    message: 'Jitsi Meeting Names:'
+                }}
+                radioGlobal={'radio'}
```
Comment (manojmalik20) [inline webapp/src/components/admin_settings/jitsi_section.tsx]: This can also be a boolean prop `isGlobal`. And if it's compulsory for a radio button to either be `inline` or `global`, then only one prop is enough.

File: webapp/src/components/InputField/index.tsx (hunk 19 above)
Comment (ayusht2810) [reply webapp/src/components/InputField/index.tsx]: type is prop in "input" component which can be 'input' or 'number' and tagType here can be "input" or "textarea"

File: webapp/src/components/admin_settings/jitsi_section.tsx (hunk 32 above)
Comment (ayusht2810) [reply webapp/src/components/admin_settings/jitsi_section.tsx]: using single prop 'isInline'

File: webapp/src/components/InputField/index.tsx [hunk 33]
```diff
@@ -1,70 +1,52 @@
...
+import React, {ReactNode} from 'react';
 
 import {InputFieldType} from 'types';
 
 type Props = {
-    heading: {
-        id: string;
-        message: string;
-    };
+    heading: ReactNode;
```
Comment (manojmalik20) [inline webapp/src/components/InputField/index.tsx]:
```suggestion
    heading: string | ReactNode;
```
Do this change everywhere where you have used only `ReactNode`.

File: webapp/src/components/InputField/index.tsx [hunk 34]
```diff
@@ -1,70 +1,52 @@
...
     tagType: string;
```
Comment (manojmalik20) [inline webapp/src/components/InputField/index.tsx]: You can remove this prop and use the `type` prop only.

File: webapp/src/components/InputField/index.tsx [hunk 35]
```diff
@@ -1,70 +1,52 @@
...
-        defaultValue?: number;
-        onChange: (e: React.ChangeEvent<InputFieldType>) => void;
-        value: string | number;
-        disabled?: boolean;
-    };
-    description: {
-        id: string;
-        message: string;
-    }
+    type?: string;
```
Comment (manojmalik20) [inline webapp/src/components/InputField/index.tsx]:
```suggestion
    type?: 'text' | 'number' | 'textarea';
```
Above is just an example. You should create an enum for this.

File: webapp/src/components/InputField/index.tsx [hunk 36]
```diff
@@ -1,70 +1,52 @@
...
+    placeholder?: string;
+    min?: number;
+    rows?: number;
+    defaultValue?: number;
```
Comment (manojmalik20) [inline webapp/src/components/InputField/index.tsx]: If the value prop is `string` | `number`. Then the `defaultValue` prop should have the same value.

File: webapp/src/components/RadioField/index.tsx [hunk 37]
```diff
@@ -1,73 +1,48 @@
...
-                                />
-                            </span>
-                        </div>
-                    </>
+        <div className='form-group'>
+            <label className='col-sm-4'>
+                {heading}
+            </label>
+            <div className='col-sm-8'>
+                <div className={`${isInline ? '' : 'radio'}`}>
```
Comment (manojmalik20) [inline webapp/src/components/RadioField/index.tsx]:
```suggestion
                <div className={`${!isInline && 'radio'}`}>
```

File: webapp/src/components/RadioField/index.tsx [hunk 38]
```diff
@@ -1,73 +1,48 @@
...
+                    {options.map((option) => (
+                        <label
+                            className={`pt-0 ${isInline ? 'radio-inline' : ''}`}
```
Comment (manojmalik20) [inline webapp/src/components/RadioField/index.tsx]:
```suggestion
                            className={`pt-0 ${isInline && 'radio-inline'}`}
```

File: webapp/src/types.ts [hunk 39]
```diff
@@ -18,3 +18,10 @@ export type GlobalState = ReduxGlobalState & {
 }
 
 export type InputFieldType = HTMLInputElement | HTMLTextAreaElement;
+
+export type RadioOptionsType = {
+    value: string;
+    checked: boolean;
+    id: string;
+    message: string;
```
Comment (manojmalik20) [inline webapp/src/types.ts]:
```suggestion
    label: string | ReactNode;
```

File: webapp/src/types.ts [hunk 40]
```diff
@@ -25,3 +25,9 @@ export type RadioOptionsType = {
     id: string;
     message: string;
 }
+
+export enum Types {
+    text = 'text',
+    number = 'number',
+    textArea = 'textarea',
```
Comment (manojmalik20) [inline webapp/src/types.ts]:
```suggestion
export enum InputTypes {
    Text = 'text',
    Number = 'number',
    Textarea = 'textarea',
```
Also, change the name of type `InputFieldType` to `InputElementType`.

File: webapp/src/types.ts (hunk 39 above)
Comment (manojmalik20) [reply webapp/src/types.ts]: You have not done this change, yet you resolved this comment. Any reason? @ayusht2810

Comment (ayusht2810) [reply webapp/src/types.ts]: I thought we need to do this on individual props only. updated now.

File: webapp/src/components/RadioField/index.tsx (hunk 24 above)
Comment (ayusht2810) [reply webapp/src/components/RadioField/index.tsx]: using single prop isInline now

File: webapp/src/components/RadioField/index.tsx (hunk 23 above)
Comment (ayusht2810) [reply webapp/src/components/RadioField/index.tsx]: not applying this. changed to single prop "isInline"

File: webapp/src/components/RadioField/index.tsx (hunk 25 above)
Comment (ayusht2810) [reply webapp/src/components/RadioField/index.tsx]: not applying this. changed to single prop "isInline"

File: webapp/src/components/admin_settings/jaas_section.tsx (hunk 31 above)
Comment (ayusht2810) [reply webapp/src/components/admin_settings/jaas_section.tsx]: reverted the change

**Review by manojmalik20 — APPROVED** [review]
